File: kotlin/src/test/kotlin/com/slothlabs/envlint/ValueTest.kt
package com.slothlabs.envlint

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds

private fun ok(r: CoercionResult): Value = (r as CoercionResult.Ok).value

private fun isErr(r: CoercionResult): Boolean = r is CoercionResult.Err

class ValueTest {
    @Test
    fun coercesScalars() {
        assertEquals(Value.Int(42), ok(coerce("42", VarType.INT)))
        assertEquals(Value.Float(3.5), ok(coerce("3.5", VarType.FLOAT)))
        assertEquals(Value.Bool(true), ok(coerce("YES", VarType.BOOL)))
        assertEquals(Value.Bool(false), ok(coerce("off", VarType.BOOL)))
        assertEquals(Value.Port(8080), ok(coerce("8080", VarType.PORT)))
    }

    @Test
    fun rejectsBadScalars() {
        assertTrue(isErr(coerce("not-int", VarType.INT)))
        assertTrue(isErr(coerce("0", VarType.PORT)))
        assertTrue(isErr(coerce("70000", VarType.PORT)))
        assertTrue(isErr(coerce("maybe", VarType.BOOL)))
    }

    @Test
    fun urlShape() {
        assertTrue(isUrl("https://example.com"))
        assertTrue(isUrl("postgres://user:pass@db:5432/app"))
        assertTrue(!isUrl("example.com"))
        assertTrue(!isUrl("://nohost"))
        assertTrue(!isUrl("1http://x"))
        assertTrue(!isUrl("http:///path"))
    }

    @Test
    fun durations() {
        assertEquals(500.milliseconds, parseDuration("500ms"))
        assertEquals(30.seconds, parseDuration("30s"))
        assertEquals(300.seconds, parseDuration("5m"))
        assertEquals(7200.seconds, parseDuration("2h"))
        assertEquals(86_400.seconds, parseDuration("1d"))
        assertEquals(10.seconds, parseDuration("10"))
        assertEquals(null, parseDuration("-1s"))
        assertEquals(null, parseDuration("abc"))
    }

    @Test
    fun durationValueDisplaysAsMillis() {
        assertEquals("30000ms", Value.Dur(30.seconds).toString())
    }
}

File: ts/test/value.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { coerce, isUrl, parseDuration } from "../src/index.ts";

test("coerces scalars", () => {
  assert.deepEqual(coerce("42", "int"), { ok: true, value: { kind: "int", value: 42 } });
  assert.deepEqual(coerce("3.5", "float"), { ok: true, value: { kind: "float", value: 3.5 } });
  assert.deepEqual(coerce("YES", "bool"), { ok: true, value: { kind: "bool", value: true } });
  assert.deepEqual(coerce("off", "bool"), { ok: true, value: { kind: "bool", value: false } });
  assert.deepEqual(coerce("8080", "port"), { ok: true, value: { kind: "port", value: 8080 } });
});

test("rejects bad scalars", () => {
  assert.equal(coerce("not-int", "int").ok, false);
  assert.equal(coerce("3.5", "int").ok, false); // floats are not ints
  assert.equal(coerce("0", "port").ok, false); // below range
  assert.equal(coerce("70000", "port").ok, false); // above range
  assert.equal(coerce("maybe", "bool").ok, false);
});

test("url shape", () => {
  assert.equal(isUrl("https://example.com"), true);
  assert.equal(isUrl("postgres://user:pass@db:5432/app"), true);
  assert.equal(isUrl("example.com"), false);
  assert.equal(isUrl("://nohost"), false);
  assert.equal(isUrl("1http://x"), false);
  assert.equal(isUrl("http:///path"), false);
});

test("durations parse to milliseconds", () => {
  assert.equal(parseDuration("500ms"), 500);
  assert.equal(parseDuration("30s"), 30_000);
  assert.equal(parseDuration("5m"), 300_000);
  assert.equal(parseDuration("2h"), 7_200_000);
  assert.equal(parseDuration("1d"), 86_400_000);
  assert.equal(parseDuration("10"), 10_000); // bare number = seconds
  assert.equal(parseDuration("-1s"), undefined);
  assert.equal(parseDuration("abc"), undefined);
});

test("string and enum carry through untyped", () => {
  assert.deepEqual(coerce("anything", "string"), {
    ok: true,
    value: { kind: "string", value: "anything" },
  });
  assert.deepEqual(coerce("debug", "enum"), {
    ok: true,
    value: { kind: "string", value: "debug" },
  });
});
